src/clj/kixi/hecuba/data/measurements/calculations.clj

```diff
@@ -242,3 +244,119 @@
   (let [{:keys [start-date end-date]} range]
     (doseq [timestamp (time/seq-dates start-date end-date (t/days 1))]
       (calculate-batch store sensor (t/minus timestamp (t/weeks 4)) (t/plus timestamp (t/days 1)) calculation-type))))
+
+(defn read-tariffs
+  "Reads profiles from the database and extracts tariffs."
+  [store entity_id]
+  (db/with-session [session (:hecuba-session store)]
+    (let [profiles (profiles/get-profiles entity_id session)]
+      (mapcat #(hash-map (:timestamp %) (-> % :profile_data :tariff)) profiles))))
+
+(defn tariff-ranges
+  "Filters timestamps from tariffs, and creates ranges of datetimes
+  that span each tariff."
+  [tariffs]
+  (let [tarrifs-count (count tariffs)]
+    (map-indexed (fn [idx item]
+                   (let [next (inc idx)]
+                     (hash-map :start (tc/to-date-time (:timestamp item))
+                               :end (if (< next tarrifs-count)
+                                      (tc/to-date-time (:timestamp (nth tariffs next)))
+                                      (time/update-timestamp (t/now) {:hour 0 :minutes 0 :seconds 0
+                                                                      :milliseconds 0}))
+                               :tariff (-> item :profile_data :tariff)))) tariffs)))
+
+(defn match-tariff
+  "Finds tariff that matches period of a measurement."
+  [m tariffs]
+  (let [t (-> m :timestamp tc/to-date-time)]
+    (first (keep #(when (time/in-range? t (:start %) (:end %)) %) (tariff-ranges tariffs)))))
```

Comment:
Can we assume that there will always be an active tariff and use min-key instead. This might simplify tariff-ranges too.

https://clojuredocs.org/clojure.core/min-key